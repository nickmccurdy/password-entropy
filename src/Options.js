import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

export default class Options extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    onChange: PropTypes.func.isRequired,
    possiblePasswords: PropTypes.func.isRequired
  }

  componentDidMount = this.change
  componentDidUpdate = this.change

  change () {
    this.props.onChange(this.props.possiblePasswords())
  }

  render () {
    return <div>{this.props.children}</div>
  }
}
