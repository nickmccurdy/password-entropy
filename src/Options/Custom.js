import { faQuestionCircle } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import { FormGroup, Input, Label } from 'reactstrap'

export default function Custom({ custom, onCustom }) {
  return (
    <FormGroup>
      <Label>
        <FontAwesomeIcon icon={faQuestionCircle} /> Possible Items
        <Input
          value={custom}
          onChange={onCustom}
          type="number"
          min="0"
          required
        />
      </Label>
    </FormGroup>
  )
}

Custom.propTypes = {
  custom: PropTypes.number.isRequired,
  onCustom: PropTypes.func.isRequired,
}
