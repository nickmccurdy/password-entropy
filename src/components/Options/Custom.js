import TextField from 'material-ui/TextField'
import Edit from 'material-ui-icons/Edit'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'

import { selector, setCustom } from '../../reducers/custom'

export function Custom({ possibleItems, setCustom }) {
  return (
    <TextField label={<div><Edit /> Possible Items</div>} value={possibleItems} onChange={setCustom} type="number" min="0" required />
  )
}

Custom.propTypes = {
  possibleItems: PropTypes.number.isRequired,
  setCustom: PropTypes.func.isRequired
}

export default connect(selector, { setCustom })(Custom)
