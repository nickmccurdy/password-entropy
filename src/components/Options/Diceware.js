import TextField from 'material-ui/TextField'
import Edit from 'material-ui/svg-icons/image/edit'
import React from 'react'

import { dicewareWords } from '../../reducers/possibleItemsSelector'

export default function Diceware() {
  return (
    <TextField
      disabled
      floatingLabelText={<div><Edit /> Possible Items</div>}
      value={dicewareWords}
    />
  )
}
