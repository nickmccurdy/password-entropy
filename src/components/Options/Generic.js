import startCase from 'lodash/startCase'
import Checkbox from 'material-ui/Checkbox'
import entries from 'object.entries'
import PropTypes from 'prop-types'
import always from 'ramda/src/always'
import map from 'ramda/src/map'
import React from 'react'
import { connect } from 'react-redux'

import { selector, toggleGeneric, toggles } from '../../reducers/generic'

export function Generic(props) {
  return (
    <div>
      {entries(toggles).map(([name, { example }]) => (
        <Checkbox
          key={name}
          name={name}
          label={<div>{ startCase(name) } <small>({ example })</small></div>}
          checked={props[name]}
          onCheck={props.toggleGeneric}
        />
      ))}
    </div>
  )
}

Generic.propTypes = {
  ...map(always(PropTypes.bool.isRequired), toggles),
  toggleGeneric: PropTypes.func.isRequired
}

export default connect(selector, { toggleGeneric })(Generic)
