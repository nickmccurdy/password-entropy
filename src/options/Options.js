import startCase from 'lodash/startCase'
import entries from 'object.entries'
import PropTypes from 'prop-types'
import pick from 'ramda/src/pick'
import React from 'react'
import Panel from 'react-bootstrap/lib/Panel'
import Tab from 'react-bootstrap/lib/Tab'
import Tabs from 'react-bootstrap/lib/Tabs'
import Icon from 'react-fa'
import { connect } from 'react-redux'

import { setOptionsKey } from '../actions'
import CustomOptions from '../options/CustomOptions'
import DicewareOptions from '../options/DicewareOptions'
import GenericOptions from '../options/GenericOptions'

const options = {
  generic: {
    Component: GenericOptions,
    icon: 'check-square'
  },
  diceware: {
    Component: DicewareOptions,
    icon: 'book'
  },
  custom: {
    Component: CustomOptions,
    icon: 'question-circle'
  }
}

export function Options({ optionsKey, setOptionsKey }) {
  return (
    <Tabs activeKey={optionsKey} onSelect={setOptionsKey} id="options">
      {entries(options).map(([name, { Component, icon }]) => (
        <Tab key={name} eventKey={name} title={<div><Icon name={icon} /> {startCase(name)}</div>}>
          <Panel><Component /></Panel>
        </Tab>
      ))}
    </Tabs>
  )
}

Options.propTypes = {
  optionsKey: PropTypes.string.isRequired,
  setOptionsKey: PropTypes.func.isRequired
}

export const mapStateToProps = pick(['optionsKey'])

export default connect(mapStateToProps, { setOptionsKey })(Options)
