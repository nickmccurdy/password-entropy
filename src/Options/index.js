import { Card, CardContent, TextField, Typography } from '@material-ui/core'
import PropTypes from 'prop-types'
import React from 'react'

import Custom from './Custom'
import Generic from './Generic'

export default function Options({
  length,
  onCustom,
  onGeneric,
  onLength,
  options,
  optionsKey
}) {
  return (
    <Card>
      <CardContent>
        <Typography variant="title" gutterBottom>
          Options
        </Typography>

        <TextField
          label="Number of Items"
          value={length}
          onChange={onLength}
          type="number"
          min="0"
          autoFocus
          fullWidth
        />

        {
          {
            Generic: (
              <Generic onGeneric={onGeneric} toggles={options.Generic} />
            ),
            Diceware: null,
            Custom: <Custom custom={options.Custom} onCustom={onCustom} />
          }[optionsKey]
        }
      </CardContent>
    </Card>
  )
}

Options.propTypes = {
  length: PropTypes.number.isRequired,
  onCustom: PropTypes.func.isRequired,
  onGeneric: PropTypes.func.isRequired,
  onLength: PropTypes.func.isRequired,
  options: PropTypes.shape({
    Custom: PropTypes.number.isRequired,
    Generic: PropTypes.shape({
      'Lowercase Letters': PropTypes.bool.isRequired,
      'Uppercase Letters': PropTypes.bool.isRequired,
      Numbers: PropTypes.bool.isRequired,
      Symbols: PropTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  optionsKey: PropTypes.string.isRequired
}
