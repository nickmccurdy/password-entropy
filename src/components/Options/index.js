import startCase from 'lodash/startCase'
import { Card, CardText } from 'material-ui/Card'
import { Tabs, Tab } from 'material-ui/Tabs'
import Book from 'material-ui/svg-icons/action/book'
import Check from 'material-ui/svg-icons/navigation/check'
import Edit from 'material-ui/svg-icons/image/edit'
import entries from 'object.entries'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'

import Custom from './Custom'
import Diceware from './Diceware'
import Generic from './Generic'
import { selector, setOptionsKey } from '../../reducers/optionsKey'

const options = {
  generic: {
    component: <Generic />,
    icon: <Check />
  },
  diceware: {
    component: <Diceware />,
    icon: <Book />
  },
  custom: {
    component: <Custom />,
    icon: <Edit />
  }
}

export function Options({ optionsKey, setOptionsKey }) {
  return (
    <Card>
      <Tabs value={optionsKey} onChange={setOptionsKey}>
        {entries(options).map(([name, { component, icon }]) => (
          <Tab key={name} icon={icon} label={startCase(name)} value={name}>
            <CardText>{component}</CardText>
          </Tab>
        ))}
      </Tabs>
    </Card>
  )
}

Options.propTypes = {
  optionsKey: PropTypes.string.isRequired,
  setOptionsKey: PropTypes.func.isRequired
}

export default connect(selector, { setOptionsKey })(Options)
