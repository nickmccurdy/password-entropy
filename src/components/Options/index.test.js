import { shallow } from 'enzyme'
import React from 'react'

import Options from '.'

test('Options', () => {
  const handleOptionsKey = jest.fn()
  const wrapper = shallow(
    <Options
      onCustom={() => {}}
      onGeneric={() => {}}
      onOptionsKey={handleOptionsKey}
      options={{
        Custom: 0,
        Generic: {
          Letters: true,
          'Capital Letters': true,
          Numbers: true,
          Symbols: true
        }
      }}
      optionsKey="Generic"
    />
  )
  wrapper
    .find('NavLink')
    .first()
    .simulate('click')
  expect(wrapper).toMatchSnapshot()
  expect(handleOptionsKey).toHaveBeenCalled()
})
