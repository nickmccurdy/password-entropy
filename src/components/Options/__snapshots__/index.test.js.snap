// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
>
  <Tabs
    initialSelectedIndex={0}
    onChange={[Function]}
    value="generic"
  >
    <Tab
      icon={<NavigationCheck />}
      label="Generic"
      value="generic"
    >
      <CardText>
        <Connect(Generic) />
      </CardText>
    </Tab>
    <Tab
      icon={<ActionBook />}
      label="Diceware"
      value="diceware"
    >
      <CardText>
        <Diceware />
      </CardText>
    </Tab>
    <Tab
      icon={<ImageEdit />}
      label="Custom"
      value="custom"
    >
      <CardText>
        <Connect(Custom) />
      </CardText>
    </Tab>
  </Tabs>
</Card>
`;
